[RCTBridge] Clean up reload notification observer

ide opened the pull request:
After the RCTBatchedBridge refactor, the `-[NSNotificationCenter addObserver]` and `[removeObserver]` calls got divided between RCTBridge and RCTBatchedBridge. This diff does two things:

 - Moved `removeObserver` out of RTCBatchedBridge and into `-[RCTBridge invalidate]`
 - Moved `addObserver` from `bindKeys` to `setUp`. This is so that `-[RCTBridge reload]` will re-add the observer after invalidating and removing the observer

cc @tadeuzagallo

tadeuzagallo commented:
Thanks, I missed that.

Any reason not to just move the `removeObserver` to the `RCTBridge`'s `dealloc`? I don't see any reason to keep adding and removing the observer since we no longer need to wait for the modules to be loaded in order to reload.

ide commented:
I put it in `-[invalidate]` so that if you were to invalidate the bridge directly (when not reloading it), it would no longer be observing the reload events. Alternatively we could observe and just not handle the notification when the bridge is invalid, and remove the observer in `-[dealloc]`. The main thing is I wanted to fix a crash I was getting about the notification being sent to an unallocated RCTBridge instance.

tadeuzagallo commented:
As I see it, removing the observer is not part of invalidating the bridge, so it should live in `-[dealloc]`. About not handling the notification when the bridge is invalid, I don't think we should worry about that. It basically just calls `-[RCTBatchedBridge invalidate]`, and the first thing it does is check if it's valid, so it should be fine.

ide commented:
Moved the `removeObserver` call to dealloc. I added an isValid check to `-[reload]` because of the dispatch_async call:

1. Bridge is invalid, receives reload notification, runs dispatch_async
2. Bridge becomes valid before the dispatched block runs
3. Block runs on main queue and invalidates the bridge again

tadeuzagallo commented:
How would the bridge become valid on 2.? What 3. could cause?

Remember that now every time you reload the bridge, you create a new BatchedBridge, so `-isValid` can never be `NO`, and then `YES` again on the same instance, so you can never invalidate the same instance twice.

ide commented:
You're right, the current code should be fine since there is no public API to make a bridge valid. My thinking is that programming defensively would make it easier to safely make changes to RCTBridge without causing extra reloads in apps that manage multiple bridges. (I want to support the use case of multiple RCTBridge instances, where the invalidation of one bridge can cause the invalidation of another with an explicit `-[invalidate]` call.)

One slightly different issue I realized is that if you reload a bridge and invalidate it, it will still reload. Shouldn't invalidating a bridge cancel all operations related to the bridge? That is -[reload] should look like:

```
- (void)reload
{
/**
 * AnyThread
 */
  typeof(self) __weak weakSelf = self;
  dispatch_async(dispatch_get_main_queue(), ^{
    typeof(self) strongSelf = weakSelf;
    if (!strongSelf || !strongSelf.isValid) {
      return;
    }
    
    [strongSelf invalidate];
    [strongSelf setUp];
  });
}
```

ide commented:
For this diff I reduced it just to the `removeObserver` call. Will make a separate PR for reloading.

tadeuzagallo commented:
That's something slightly more complicated, if we change it in a way that invalidating the bridge won't allow to reload anymore, then we should probably change the condition that determines whether `RCTBridge` is valid or not, and `-reload` probably shouldn't call `-invalidate` anymore.

I'll merge it as is for now though, thanks!
